### PR TITLE
PARQUET-1935: [C++] Fix bug in WriteBatchSpaced

### PR DESCRIPTION
**Pull request by emkornfield**
The arrow code path doesn't call this method of not-null columns with no null parents, so this wasn't caught in arrow tests.

The only time we can pass through counts is if definition level happens to be zero.

Also fix some potential UBSan issues with adding offset to values.

### CODE REVIEW

File: cpp/src/parquet/column_writer.cc [hunk 1]
```diff
@@ -1190,7 +1192,7 @@ class TypedColumnWriterImpl : public ColumnWriterImpl, public TypedColumnWriter<
                                   int64_t* out_spaced_values_to_write,
                                   int64_t* null_count) {
     if (bits_buffer_ == nullptr) {
-      if (!level_info_.HasNullableValues()) {
+      if (level_info_.def_level == 0) {
```
Comment (pitrou) [inline cpp/src/parquet/column_writer.cc:1195]: Interesting. Why is a special case needed for this? Would you like to add a comment?

Comment (pitrou) [reply cpp/src/parquet/column_writer.cc:1195]: Do you also want to check `def_levels` for null/non-null-ness? (debug mode check perhaps?)

**Review by pitrou — COMMENTED** [review]
Thank you for doing this. Just one question.

File: cpp/src/parquet/column_writer.cc (hunk 1 above)
Comment (emkornfield) [reply cpp/src/parquet/column_writer.cc:1199]: I added comments and a check.  I document the method to make it clearer that it always updates the output parameters as well.

**Comment by pitrou** [comment]
Rebased.

**Review by pitrou — APPROVED** [review]
+1, will merge if CI is green

**Pull request closed without merging**